### PR TITLE
Add MCL core class

### PR DESCRIPTION
**Pull request by serraramiro1**
### Proposed changes

Addresses #424 

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [x] 🚀 Feature (change which adds functionality)
- [ ] 📚 Documentation (change which fixes or extends documentation)

### Checklist

_Put an `x` in the boxes that apply. This is simply a reminder of what we will require before merging your code._

- [x] Lint and unit tests (if any) pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### CODE REVIEW

File: beluga/include/beluga/algorithm/mcl_core.hpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+ * \tparam WeightT Type to represent a weight of a particle.
+ * \tparam ParticleType Full particle type, containing state, weight and possibly
+ * other information .
+ * \tparam ExecutionPolicy Execution policy for particles processing.
+ */
+template <
+    class MotionModel,
+    class SensorModel,
+    typename WeightT = beluga::Weight,
+    class ParticleType = std::tuple<typename SensorModel::state_type, WeightT>,
```
Comment (hidmic) [inline beluga/include/beluga/algorithm/mcl_core.hpp:66]: @serraramiro1 @nahueespinosa meta: moving forward, this looks unnecessarily restrictive. `SensorModel` could work with multiple state types. Same can be said about `MotionModel`.

**Review by hidmic — COMMENTED** [review]
First pass

File: beluga/include/beluga/algorithm/mcl_core.hpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    class ExecutionPolicy = std::execution::sequenced_policy>
+class Mcl {
```
[the remark this answers is not in the document]
Comment (nahueespinosa) [reply beluga/include/beluga/algorithm/mcl_core.hpp:68]: I remember having a quick talk about using a CRTP base class for this. It seemed like there wasn't much to reuse, but looking at the code, there might be enough to justify a base class?

[the remark this answers is not in the document]
Comment (hidmic) [reply beluga/include/beluga/algorithm/mcl_core.hpp:68]: There is substantial overlap, though I'm glancing over the details of how we would reuse the code.

File: beluga/include/beluga/algorithm/mcl_core.hpp (hunk 1 above)
Comment (hidmic) [reply beluga/include/beluga/algorithm/mcl_core.hpp:66]: I'm wondering if we should couple the particle type and the models like these. Models could work with multiple types at the same time (e.g. through generic lambdas), so sparing the user from explicitly setting the particle state type on class instantiation seems more trouble than it's worth.

**Pull request closed without merging**